repose-aggregator/services/atom-feed/impl/src/main/scala/org/openrepose/nodeservice/atomfeed/impl/AtomEntryStreamBuilder.scala

```diff
@@ -0,0 +1,105 @@
+    * making a request to the Atom feed, assign the result of this function to a lazy variable like so:
+    * lazy val myFeed = build("http://my.feed.url", MyAuthenticatedRequestFactory)
+    *
+    * @param baseFeedUrl a static locator pointing to the "head" of an Atom feed (e.g., the subscription document)
+    * @param authenticator a URL processor which authenticates connections
+    * @param authenticationTimeLimit a maximum [[Duration]] to wait on authentication before considering authentication
+    *                                a failure
+    * @return a Scala [[scala.collection.immutable.Stream]] which will yield Atom entries in an Atom feed located at the
+    *         provided baseFeedUrl. The entries will be yielded in the order in which they are read from the feed XML,
+    *         from top to bottom.
+    * @throws AuthenticationException when the authenticator fails to authenticate a request
+    */
+  def build(baseFeedUrl: URL,
+            reposeVersion: String,
+            authenticator: Option[AuthenticatedRequestFactory] = None,
+            authenticationTimeLimit: Duration = 1 second): Stream[Entry] = {
+    val requestId = java.util.UUID.randomUUID().toString
+    MDC.put(TracingKey.TRACING_KEY, requestId)
+    buildR(baseFeedUrl, reposeVersion, authenticator, authenticationTimeLimit, requestId)
+  }
+
+  private def buildR(baseFeedUrl: URL,
+                     reposeVersion: String,
+                     authenticator: Option[AuthenticatedRequestFactory] = None,
+                     authenticationTimeLimit: Duration = 1 second,
+                     requestId: String): Stream[Entry] = {
+    val baseFeedConnection = baseFeedUrl.openConnection()
+
+    val authenticatedConnection = authenticator match {
+      case Some(arf) =>
+        val connectionFuture = Future(arf.authenticateRequest(baseFeedConnection, requestId, reposeVersion))
+        Option(Await.result(connectionFuture, authenticationTimeLimit))
+      case None =>
+        Some(baseFeedConnection)
+    }
+
+    authenticatedConnection match {
+      case Some(urlConnection) =>
+        val tracingHeader = TracingHeaderHelper.createTracingHeader(requestId, "1.1 Repose (Repose/" + reposeVersion + ")", None)
+        urlConnection.setRequestProperty(CommonHttpHeader.TRACE_GUID.toString, tracingHeader)
+
+        val feedInputStream = urlConnection.getInputStream
+        val feed = parser.parse[Feed](feedInputStream).getRoot
+        feedInputStream.close()
+
+        feed.getLinks.find(link => link.getRel.equals("next")) match {
+          case Some(nextPageLink) =>
+            feed.getEntries.toStream #::: buildR(nextPageLink.getResolvedHref.toURL, reposeVersion, authenticator, authenticationTimeLimit, requestId)
+          case None =>
+            feed.getEntries.toStream
+        }
+      case None =>
+        throw AuthenticationException
```

Comment:
Right, I started going into the case in the comment above. Basically...

If we get an unauthorized response from the Atom Feed service AFTER the first page has already been read, then we will send duplicate events. All other flows should work.

The reason is that the high water mark is only being set after all new entries have been read and sent to listeners. If we fail at any point, the high water mark is not reset. However, we send new entries as soon as possible after reading a page. So we could send some entries but never update the high water mark. That happens when a page after the first page cannot be retrieved.

So the only solution I can think of that works is buffering events until we've finished reading all of the new entries, then shooting the buffered events off to the listeners. That way, if we fail at any point, we will send no events and not update the high water mark.